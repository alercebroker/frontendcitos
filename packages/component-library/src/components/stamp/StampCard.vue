<template>
  <q-card>
    <q-card-section style="padding: 0" v-if="!hideTools">
      <div class="row justify-center">
        <div class="col-6">
          <q-select
            v-model="selected"
            filled
            dense
            :options="props.detections"
            :option-label="formatLabel"
            label="Date"
            map-options
            stack-label
            @update:model-value="changeSelect"
          />
        </div>
        <div class="col-2 q-mt-xs">
          <q-btn
            round
            style="background: white; color: black"
            size="xs"
            class="q-pa-sm q-mr-sm"
            icon="arrow_left"
            @click="previousDetection"
          >
          </q-btn>
          <q-btn
            round
            style="background: white; color: black"
            size="xs"
            class="q-pa-sm"
            icon="arrow_right"
            @click="nextDetection"
          >
          </q-btn>
        </div>
        <div class="col-1">
          <q-btn size="md" color="primary" @click="onAvroClick">AVRO</q-btn>
        </div>
        <div class="col-1">
          <q-icon name="error" size="2rem">
            <q-tooltip max-width="20rem">
              AVRO is the format in which we receive alerts from ZTF. Clicking
              the AVRO button will show the full information contained in the
              avro file for the selected detection.
            </q-tooltip>
          </q-icon>
        </div>
      </div>
    </q-card-section>
    <q-separator></q-separator>
    <q-card-section style="padding: 0 5%">
      <div class="row justify-center">
        <div class="col-4">
          <generic-card
            title="Science"
            :viewtype="data.activeTool"
            :candid="selected.candid"
            :image-url="stampImage('science')"
            :objectId="objectId"
          />
        </div>
        <div class="col-4">
          <generic-card
            title="Template"
            :viewtype="data.activeTool"
            :candid="selected.candid"
            :image-url="stampImage('template')"
            :objectId="objectId"
          />
        </div>
        <div class="col-4">
          <generic-card
            title="Difference"
            :viewtype="data.activeTool"
            :candid="selected.candid"
            :image-url="stampImage('difference')"
            :objectId="objectId"
          />
        </div>
      </div>
    </q-card-section>
    <q-separator></q-separator>
    <q-card-section v-if="!hideTools">
      <div class="row justify-center">
        <q-icon
          v-for="tool in tools"
          :key="tool.id"
          :name="tool.icon"
          size="2rem"
          class="q-mr-sm"
          @click="selectTool(tool.id)"
        ></q-icon>
        <q-icon name="question_mark" size="2rem">
          <q-tooltip max-width="20rem">
            Use the tool icons to change between <strong>zoom</strong> or
            <strong>crosshair</strong> modes
          </q-tooltip>
        </q-icon>
      </div>
    </q-card-section>
  </q-card>
</template>

<script lang="ts" setup>
import { ref, reactive } from "vue";
import { stampUrl } from "../../utils/urls";
import { jdToDate } from "../../utils/dates";
import { customIteratorFactory } from "../../utils/array";
import GenericCard from "./subcomponents/GenericCard.vue";

const tools = [
  { id: "zoom", text: "Zoom", icon: "zoom_in" },
  { id: "crosshair", text: "Crosshair", icon: "location_searching" },
];

const props = defineProps({
  detections: { type: Array, required: true },
  imageServiceUrl: { type: String, required: true },
  objectId: { type: String, required: true },
  hideTools: { type: Boolean, default: false },
});

const emit = defineEmits(["selectDetection", "toggleFullscreen", "avroClick"]);

const data = reactive({
  isFullscreen: false,
  selectedIndex: 0,
  activeTool: "crosshair",
});
const selected: any = ref(props.detections[data.selectedIndex]);
const iterator = customIteratorFactory(props.detections, data.selectedIndex);

function nextDetection() {
  iterator.next();
  data.selectedIndex = iterator.current();
  selected.value = props.detections[data.selectedIndex];
}

function previousDetection() {
  iterator.prev();
  data.selectedIndex = iterator.current();
  selected.value = props.detections[data.selectedIndex];
}

function changeSelect(element: any) {
  data.selectedIndex = iterator.moveTo(element);
}

function selectTool(toolId: string) {
  data.activeTool = toolId;
}

function onAvroClick() {
  emit("avroClick", selected);
}

function stampImage(type: string) {
  return async (candid: string, format: string) =>
    await stampUrl({
      baseUrl: props.imageServiceUrl,
      candid,
      type,
      format,
      surveyId: selected.value.tid,
    });
}

function formatLabel(detection: any) {
  return jdToDate(detection.mjd).toUTCString().slice(0, -3) + "UT";
}
</script>

<style scoped>
.col-6,
.col-1,
.col-2,
.col-3 {
  padding: 1rem;
}

.button-panel {
  display: inline-block;
  padding: 0 auto;
  justify-content: space-around;
}

.q-icon:hover {
  cursor: pointer;
  color: lightgray;
}
</style>
