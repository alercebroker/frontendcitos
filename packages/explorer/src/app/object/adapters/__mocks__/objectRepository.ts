import type {
  ObjectEntity,
  ObjectListFilters,
} from "@/domain/objects/entities";
import type { ObjectRepository } from "@/domain/objects/ports";
import {
  HttpError,
  ParseError,
  type PaginatedListEntity,
} from "@alercebroker/http-client/build/main/types";
import { err, ok, type Result } from "neverthrow";

let testType: string;

export function __setTestType(tt: string) {
  testType = tt;
}

export const objectRepository: ObjectRepository = {
  getObject: (): Promise<Result<ObjectEntity, HttpError | ParseError>> => {
    throw new Error("Not Implemented");
  },
  getObjects: (
    _filters: ObjectListFilters
  ): Promise<
    Result<PaginatedListEntity<ObjectEntity>, ParseError | HttpError>
  > => {
    if (testType === "success") {
      return new Promise((resolve) => {
        resolve(
          ok({
            total: 1,
            page: 1,
            next: 2,
            hasNext: false,
            prev: 0,
            hasPrev: false,
            items: [
              {
                aid: "aid",
                ra: 1,
                dec: 1,
                firstmjd: 1,
                lastmjd: 1,
                firstGreg: "utc",
                lastGreg: "utc",
                raHms: "hms",
                decHms: "hms",
              },
            ],
          })
        );
      });
    }
    if (testType === "clientError") {
      return new Promise((resolve) => {
        resolve(err(HttpError.fromStatus(400)));
      });
    }
    if (testType === "serverError") {
      return new Promise((resolve) => {
        resolve(err(HttpError.fromStatus(500)));
      });
    }
    if (testType === "parseError") {
      return new Promise((resolve) => {
        resolve(err(new ParseError("Parse Error")));
      });
    }
    return new Promise((resolve) => {
      resolve(err(new Error("Other cases not implemented")));
    });
  },
};
