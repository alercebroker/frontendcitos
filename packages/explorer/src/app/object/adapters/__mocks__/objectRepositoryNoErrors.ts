import type {
  ObjectEntity,
  ObjectListFilters,
  SingleObjectEntity,
  LightCurveEntity,
} from "@/domain/objects/entities";
import type { ObjectRepository } from "@/domain/objects/ports";
import type { HttpError, ParseError, PaginatedListEntity } from "@alercebroker/http-client/build/main/types/types";
import { err, ok, type Result } from "neverthrow";


export const objectRepository: ObjectRepository = {
  getObject: (
    id: string
  ): Promise<Result<SingleObjectEntity, HttpError | ParseError>> => {
    return new Promise((resolve) => {
      resolve(
        ok({
          object_basic_info: {
            aid: "aid",
            oid: ["oid1", "oid2"],
            ra: 1,
            dec: 1,
            firstmjd: 1,
            lastmjd: 1,
            firstGreg: "utc",
            lastGreg: "utc",
            raHms: "hms",
            decHms: "hms",
            ndet: 1,
          },
          magstats: [
            {
              fid: 1,
              ndet: 116,
              magmean: 18.709673,
              magmedian: 18.653234,
              magmax: 19.837292,
              magmin: 17.859047,
              magsigma: 0.4470002,
              maglast: 18.36436,
              magfirst: 18.229092,
              firstmjd: 58366.44,
              lastmjd: 59542.28,
              ingestion_step: "0.4333",
            },
            {
              fid: 2,
              ndet: 50,
              magmean: 17.327902,
              magmedian: 17.32121,
              magmax: 18.767,
              magmin: 16.580177,
              magsigma: 0.4340397,
              maglast: 18.5926,
              magfirst: 17.64601,
              firstmjd: 58366.508,
              lastmjd: 59530.367,
              ingestion_step: "0.1234",
            }
          ],
          probabilities: [
            {
              classifier_name: "stamp_classifier",
              classifier_version: "stamp_classifier_1.0.0",
              class_name: "AGN",
              probability: 0.086826704,
              ranking: 2
            },
            {
              classifier_name: "stamp_classifier",
              classifier_version: "stamp_classifier_1.0.0",
              class_name: "asteroid",
              probability: 0.059574142,
              ranking: 4
            },
            {
              classifier_name: "stamp_classifier",
              classifier_version: "stamp_classifier_1.0.0",
              class_name: "bogus",
              probability: 0.07940478,
              ranking: 3
            },
            {
              classifier_name: "stamp_classifier",
              classifier_version: "stamp_classifier_1.0.0",
              class_name: "SN",
              probability: 0.055563178,
              ranking: 5
            },
            {
              classifier_name: "stamp_classifier",
              classifier_version: "stamp_classifier_1.0.0",
              class_name: "VS",
              probability: 0.7186312,
              ranking: 1
            },
          ],
        })
      );
    });
  },
  getLightCurve: (
    id: string
  ): Promise<Result<LightCurveEntity, HttpError | ParseError>> => {
    return new Promise((resolve) => {
      resolve(
        ok({
          detections: [
            {
              aid: "AL22kyjplrvzrqwhg",
              oid: "ZTF18achqnnp",
              tid: "ZTF",
              mjd: 59813.449780100025,
              candid: "2059449780215010011",
              fid: 2,
              isdiffpos: -1,
              mag: 18.094682693481445,
              e_mag: 0.12231238931417465,
              ra: 69.3842877,
              dec: 15.2353628,
              rb: 0.45428571105003357,
              rbversion: "t17_f5_c3",
              has_stamp: true,
              corrected: true,
              step_id_corr: "dev",
              parent_candid: "nan"
            },
            {
              aid: "AL22kyjplrvzrqwhg",
              oid: "ZTF18achqnnp",
              tid: "ZTF",
              mjd: 59814.448356499895,
              candid: "2060448356015010001",
              fid: 1,
              isdiffpos: -1,
              mag: 19.929550170898438,
              e_mag: 0.21395735442638397,
              ra: 69.3844655,
              dec: 15.2354378,
              rb: 0.949999988079071,
              rbversion: "t17_f5_c3",
              has_stamp: true,
              corrected: true,
              step_id_corr: "dev",
              parent_candid: "nan"
            },
          ],
          non_detections: [
            {
              tid: "ZTF",
              mjd: 59783.47290509986,
              fid: 2,
              diffmaglim: 19.330299377441406
            },
            {
              tid: "ZTF",
              mjd: 59783.47538190009,
              fid: 2,
              diffmaglim: 19.388599395751953
            },
          ],
        })
      );
    });
  },  
  getObjects: (
    _filters: ObjectListFilters
  ): Promise<
    Result<PaginatedListEntity<ObjectEntity>, ParseError | HttpError>
  > => {
    return new Promise((resolve) => {
      resolve(
        ok({
          total: 1,
          page: 1,
          next: 2,
          hasNext: false,
          prev: 0,
          hasPrev: false,
          items: [
            {
              aid: "aid",
              ra: 1,
              dec: 1,
              ndet: 1,
              firstmjd: 1,
              lastmjd: 1,
              firstGreg: "utc",
              lastGreg: "utc",
              raHms: "hms",
              decHms: "hms",
            },
          ],
        })
      );
    });
  },
};
