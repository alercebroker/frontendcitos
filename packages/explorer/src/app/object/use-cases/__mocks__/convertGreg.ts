import type { Callbacks, Command } from "@/common/use-case";
import type { GregToMjd } from "@/domain/objects/ports";

let testType: string;

export const __setTestType = (tt: string) => {
  testType = tt;
};

export const convertGregUseCase = (_gregToMjd: GregToMjd): Command => ({
  execute: (callbacks: Callbacks, _payload: string) => {
    if (testType === "success") {
      callbacks.handleSuccess(49730);
    }
  },
});
