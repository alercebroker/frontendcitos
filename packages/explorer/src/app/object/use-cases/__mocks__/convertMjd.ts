import type { Callbacks, Command } from "@/common/use-case";
import type { MjdToGreg } from "@/domain/objects/ports";

let testType: string;

export const __setTestType = (tt: string) => {
  testType = tt;
};

export const convertMjdUseCase = (_mjdToGregFn: MjdToGreg): Command => ({
  execute: (callbacks: Callbacks, _payload: number) => {
    if (testType === "success") {
      callbacks.handleSuccess(new Date("1995-01-13").toString());
    }
  },
});
