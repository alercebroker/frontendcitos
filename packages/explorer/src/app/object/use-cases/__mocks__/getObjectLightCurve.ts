import type { Callbacks, Command } from "@/common/use-case";
import type { ObjectRepository } from "@/domain/objects/ports";

let testType: string;

export const __setTestType = (tt: string) => {
  testType = tt;
};

export const getObjectLightCurveUseCase = (
  _repository: ObjectRepository
): Command => ({
  execute: (
    callbacks: Callbacks,
    payload: string,
  ): void => {
    if (testType === "success") {
      callbacks.handleSuccess({
        detections: [
          {
            aid: "AL22kyjplrvzrqwhg",
            oid: "ZTF18achqnnp",
            tid: "ZTF",
            mjd: 59813.449780100025,
            candid: "2059449780215010011",
            fid: 2,
            isdiffpos: -1,
            mag: 18.094682693481445,
            e_mag: 0.12231238931417465,
            ra: 69.3842877,
            dec: 15.2353628,
            rb: 0.45428571105003357,
            rbversion: "t17_f5_c3",
            has_stamp: true,
            corrected: true,
            step_id_corr: "dev",
            parent_candid: "nan"
          },
        ],
        non_detections: [
          {
            tid: "ZTF",
            mjd: 59783.47290509986,
            fid: 2,
            diffmaglim: 19.330299377441406
          },
        ],
      });
    }
  },
});