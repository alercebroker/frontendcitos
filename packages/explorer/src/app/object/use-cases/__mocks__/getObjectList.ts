import type { Callbacks, Command } from "@/common/use-case";
import type { ObjectRepository } from "@/domain/objects/ports";

let testType: string;

export const __setTestType = (tt: string) => {
  testType = tt;
};

export const getObjectListUseCase = (
  _repository: ObjectRepository
): Command => ({
  execute: <ObjectListFilters>(
    callbacks: Callbacks,
    _: ObjectListFilters
  ): void => {
    if (testType === "success") {
      callbacks.handleSuccess({
        total: 1,
        page: 1,
        next: 2,
        hasNext: false,
        prev: 0,
        hasPrev: false,
        items: [
          {
            aid: "aid",
            ra: 1,
            dec: 1,
            firstmjd: 1,
            lastmjd: 1,
            firstGreg: "utc",
            lastGreg: "utc",
            raHms: "hms",
          },
        ],
      });
    }
  },
});
