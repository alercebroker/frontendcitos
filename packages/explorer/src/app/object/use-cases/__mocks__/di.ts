import { mjdToGreg } from "@/domain/objects/entities";
import { objectRepository } from "../../adapters/objectRepository";
import { convertMjdUseCase } from "./convertMjd";
import { getObjectListUseCase } from "./getObjectList";

const getObjectListUseCaseResolved = getObjectListUseCase(objectRepository);
const convertMjdUseCaseResolved = convertMjdUseCase(mjdToGreg);

export {
  getObjectListUseCaseResolved as getObjectListUseCase,
  convertMjdUseCaseResolved as convertMjdUseCase,
};
