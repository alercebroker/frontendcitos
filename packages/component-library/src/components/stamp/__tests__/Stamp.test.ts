/**
 * @jest-environment jsdom
 */

import { mount } from "@vue/test-utils";
import StampCard from "../StampCard.vue";
import { detections } from "./__mocks__";

test("Stamps snapshot", () => {
  const stamps = mount(StampCard as any, {
    props: {
      detections,
      imageServiceUrl: "http://avro.alerce.online/get_stamp",
      objectId: "ZTF20aaelulu",
    },
  });
  expect(stamps.element).toMatchSnapshot();
});
