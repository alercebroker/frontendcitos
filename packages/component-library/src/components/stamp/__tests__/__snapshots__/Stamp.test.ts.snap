// Vitest Snapshot v1

exports[`Stamps snapshot > snap 1`] = `
<q-card-stub
  bordered="false"
  data-v-b5e15eb2=""
  flat="false"
  square="false"
  tag="div"
/>
`;
